segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/TableBasedMetadataStore.java

```diff
@@ -0,0 +1,172 @@
+    TableStore tableStore;
+    String tableName;
+    Duration timeout = Duration.ofSeconds(1L);
+    AtomicBoolean isTableInitialized = new AtomicBoolean(false);
+
+    public TableBasedMetadataStore(String tableName, TableStore tableStore) {
+        this.tableStore = Preconditions.checkNotNull(tableStore, "tableStore");
+        this.tableName = Preconditions.checkNotNull(tableName, "tableName");
+    }
+
+    @Override
+    protected TransactionData read( String key) throws StorageMetadataException {
+        ensureInitialized();
+        List<ArrayView> keys = new ArrayList<>();
+        keys.add(new ByteArraySegment(key.getBytes()));
+        try {
+            List<TableEntry> retValue = this.tableStore.get(tableName, keys, timeout).get();
+            if (retValue.size() == 1) {
+                TableEntry entry = retValue.get(0);
+                if (null != entry) {
+                    val arr = entry.getValue();
+                    ObjectInputStream input = new ObjectInputStream(new ByteArrayInputStream(arr.array(), arr.arrayOffset(), arr.getLength()));
+                    TransactionData txnData = (TransactionData) input.readObject();
+                    txnData.setDbObject(entry.getKey().getVersion());
+                    return txnData;
+                }
+            }
+        } catch (IllegalStateException e) {
+            throw e;
+        } catch (Exception e) {
+            throw new StorageMetadataException("Error while reading", e);
+        }
+
+        return TransactionData.builder()
+                .key(key)
+                .persisted(true)
+                .dbObject(TableKey.NOT_EXISTS)
+                .build();
+    }
+
+    @Override
+    protected void writeAll(Collection<TransactionData> dataList) throws StorageMetadataException {
+        ensureInitialized();
+        List<TableEntry> toUpdate = new ArrayList<>();
+        HashMap<TableEntry, TransactionData> entryToTxnDataMap = new HashMap<TableEntry, TransactionData>();
+        HashMap<TableKey, TransactionData> deletedKeyToTxnDataMap = new HashMap<TableKey, TransactionData>();
+        List<TableKey> keysToDelete = new ArrayList<>();
+        try {
+            for (TransactionData txnData : dataList) {
+                Preconditions.checkState(null != txnData.getDbObject());
+
+                long version = ((Long) txnData.getDbObject()).longValue();
+                if (null == txnData.getValue()) {
+                    //version = TableKey.NO_VERSION;
+                    val toDelete = TableKey.versioned(new ByteArraySegment(txnData.getKey().getBytes()),
+                            TableKey.NO_VERSION);
+                    keysToDelete.add(toDelete);
+                    deletedKeyToTxnDataMap.put(toDelete, txnData);
+                }
+
+                ByteArrayOutputStream bos = new ByteArrayOutputStream();
+                ObjectOutputStream out = new ObjectOutputStream(bos);
```

Comment:
Why don't we have a better serialization mechanism ?
Java serialization is heavy and it could lead to security issues.

Reply:
Fixed in part 1